components/application_manager/opapp.cpp

```diff
@@ -0,0 +1,131 @@
+#include "opapp.h"
+#include "log.h"
+
+#define COUNT_DOWN_TIMEOUT 60000
+
+static std::string opAppStateToString(const App::E_APP_STATE &state);
+
+/**
+ * Create opapp from url.
+ * 
+ * @throws std::runtime_error
+ */
+OpApp::OpApp(const std::string &url, std::shared_ptr<OpApp::SessionCallback> sessionCallback)
+    : App(url, sessionCallback)
+{ 
+    m_state = BACKGROUND_STATE; // ETSI TS 103 606 V1.2.1 (2024-03) page 36
+}
+
+/**
+ * Create opapp from Ait description.
+ * 
+ * @throws std::runtime_error
+ */
+OpApp::OpApp(const Ait::S_AIT_APP_DESC &desc, bool isNetworkAvailable, std::shared_ptr<OpApp::SessionCallback> sessionCallback)
+    : App(
+        desc,
+        Utils::MakeInvalidDvbTriplet(),
+        isNetworkAvailable,
+        "",
+        true,
+        false,
+        sessionCallback
+    )
+{ 
+    m_state = BACKGROUND_STATE; // ETSI TS 103 606 V1.2.1 (2024-03) page 36
+}
+
+/**
+ * Create opapp from url and inherit another opapp's state (ETSI TS 103 606 V1.2.1 (2024-03) 6.3.3.1).
+ * 
+ * @throws std::runtime_error
+ */
+OpApp::OpApp(const OpApp &other, const std::string &url)
+    : OpApp(url, other.m_sessionCallback)
+{
+    m_state = other.GetState();
+    if (!other.m_countdown.isStopped())
+    {
+        m_countdown.start(other.m_countdown.remaining());
+    }
+}
+
+void OpApp::SetState(const E_APP_STATE &state)
+{
+    if (state != OVERLAID && CanTransitionToState(state))
+    {
+        if (state != m_state)
+        {
+            std::string previous = opAppStateToString(m_state);
+            std::string next = opAppStateToString(state);
+            m_state = state;
+            static_cast<SessionCallback*>(m_sessionCallback.get())->DispatchOperatorApplicationStateChange(GetId(), previous, next);
+            
+            if (state == BACKGROUND_STATE)
+            {
+                m_sessionCallback->HideApplication(GetId());
+            }
+            else
+            {
+                m_sessionCallback->ShowApplication(GetId());
+            }
+        }
+        if ((state | TRANSIENT_STATE) != 0)
+        {
+            m_countdown.start(std::chrono::milliseconds(COUNT_DOWN_TIMEOUT));
+        }
+        else
+        {
+            m_countdown.stop();
+        }
+    }
+    else
+    {
+        LOG(LOG_INFO, "Invalid state transition: %d -> %d", m_state, state);
+    }
+}
+
+bool OpApp::TransitionToBroadcastRelated()
+{
+    // ETSI TS 103 606 V1.2.1 (2024-03) 6.3.3.2 Note 2
+    if (m_state == FOREGROUND_STATE)
+    {
+        return App::TransitionToBroadcastRelated();
+    }
+    return false;
+}
+
+bool OpApp::CanTransitionToState(const E_APP_STATE &state)
```

Comment:
Sorry but I can't quite get my head around this logic. If state equals current m_state, return true? Is that correct?

The logic is confusing and I think a switch statement is not the right choice here. A better choice is if-else.

Reply:
This bit of logic needs a unit-test to exercise all paths. Can you create one?

Reply:
The logic behind this lies in SetState() method and is designed like this in order to re/start or stop the countdown timer depending of the state.
According to spec, _If the operator application is already in the transient state and makes a call to the opAppRequestTransient method and one of the requirements for the call to be successful listed above is met, then the call shall be successful and the 60 second countdown timer shall be restarted from the beginning._
That said, transitions from a given state to the same state, are valid.

Reply:
OK, removing bitmask. With regards to the reference on the spec, I'd rather keep it as is than pasting large excepts from the spec

Reply:
It's ok to have the references but an explanation is also required - I spent more than 5 minutes looking for that spec and the relevant section.